File: example-objc/example-objc/ExampleHostViewController.m
//
//  ExampleHostViewController.m
//  example-objc
//
//  Created by Tadashi Wakayanagi on 2019/11/18.
//  Copyright © 2019 PAY, Inc. All rights reserved.
//

#import "ExampleHostViewController.h"
#import "ColorTheme.h"
#import "SampleService.h"
#import "UIViewController+Alert.h"
@import PAYJP;

@interface ExampleHostViewController ()

@end

@implementation ExampleHostViewController

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
  [tableView deselectRowAtIndexPath:indexPath animated:true];

  if (indexPath.row == 3) {
    // customize card form
    //        UIColor *color = RGB(0, 122, 255);
    //        PAYCardFormStyle *style = [[PAYCardFormStyle alloc] initWithLabelTextColor:color
    //                                                                    inputTextColor:color
    //                                                                    errorTextColor:nil
    //                                                                         tintColor:color
    //                                                         inputFieldBackgroundColor:nil
    //                                                                 submitButtonColor:color];
    // push
    PAYCardFormViewController *cardFormVc = [PAYCardFormViewController
        createCardFormViewControllerWithStyle:PAYCardFormStyle.defaultStyle
                                     tenantId:nil];
    cardFormVc.delegate = self;
    [self.navigationController pushViewController:cardFormVc animated:YES];

    // modal
    //                PAYCardFormViewController *cardFormVc =
    //                    [PAYCardFormViewController
    //                    createCardFormViewControllerWithStyle:PAYCardFormStyle.defaultStyle
    //                                                                            tenantId:nil];
    //                cardFormVc.delegate = self;
    //                UINavigationController *naviVc =
    //                    [UINavigationController.new initWithRootViewController:cardFormVc];
    //                naviVc.presentationController.delegate = cardFormVc;
    //                [self presentViewController:naviVc animated:true completion:nil];
  }
}

#pragma MARK : PAYCardFormViewControllerDelegate

- (void)cardFormViewController:(PAYCardFormViewController *_Nonnull)_
               didCompleteWith:(enum CardFormResult)result {
  __weak typeof(self) wself = self;

  switch (result) {
    case CardFormResultCancel:
      NSLog(@"CardFormResultCancel");
      break;
    case CardFormResultSuccess:
      NSLog(@"CardFormResultSuccess");
      dispatch_async(dispatch_get_main_queue(), ^{
        // pop
        [wself.navigationController popViewControllerAnimated:YES];

        // dismiss
        //                  [wself.navigationController dismissViewControllerAnimated:YES
        //                  completion:nil];
      });
      break;
  }
}

- (void)cardFormViewController:(PAYCardFormViewController *)_
                   didProduced:(PAYToken *)token
             completionHandler:(void (^)(NSError *_Nullable))completionHandler {
  NSLog(@"token = %@", [self displayToken:token]);

  // サーバにトークンを送信
  SampleService *service = [SampleService sharedService];
  [service saveCardWithToken:token.identifer
                  completion:^(NSError *error) {
                    if (error != nil) {
                      NSLog(@"Failed save card. error = %@", error);
                      completionHandler(error);
                    } else {
                      NSLog(@"Success save card. token = %@", [self displayToken:token]);
                      completionHandler(nil);
                    }
                  }];
}

@end

File: example-swift/example-swift/ExampleHostViewController.swift
//
//  ExampleHostViewController.swift
//  example-swift
//
//  Created by Tadashi Wakayanagi on 2019/11/19.
//

import UIKit
import PAYJP

class ExampleHostViewController: UITableViewController {

    private var token: Token?

    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: true)

        if indexPath.row == 3 {
            // customize card form
            //            let color = UIColor(0, 122, 255)
            //            let style = FormStyle(
            //                labelTextColor: color,
            //                inputTextColor: color,
            //                tintColor: color)

            let cardFormVc = CardFormViewController.createCardFormViewController(delegate: self)
            // push
            self.navigationController?.pushViewController(cardFormVc, animated: true)

            // modal
            //            let naviVc = UINavigationController(rootViewController: cardFormVc)
            //            naviVc.presentationController?.delegate = cardFormVc
            //            self.present(naviVc, animated: true, completion: nil)
        }
    }
}

extension ExampleHostViewController: CardFormViewControllerDelegate {

    func cardFormViewController(_: CardFormViewController, didCompleteWith result: CardFormResult) {
        switch result {
        case .cancel:
            print("CardFormResult.cancel")
        case .success:
            print("CardFormResult.success")
            DispatchQueue.main.async { [weak self] in
                guard let self = self else { return }
                // pop
                self.navigationController?.popViewController(animated: true)
                if let token = self.token {
                    self.showToken(token: token)
                }

                // dismiss
                //                                self?.dismiss(animated: true, completion: nil)
            }
        }
    }

    func cardFormViewController(_: CardFormViewController,
                                didProduced token: Token,
                                completionHandler: @escaping (Error?) -> Void) {
        print("token = \(token.display)")
        self.token = token

        // サーバにトークンを送信
        SampleService.shared.saveCard(withToken: token.identifer) { (error) in
            if let error = error {
                print("Failed save card. error = \(error)")
                completionHandler(error)
            } else {
                print("Success save card.")
                completionHandler(nil)
            }
        }
    }
}
